consensus/src/creation/mod.rs

```diff
@@ -104,69 +139,95 @@ async fn create_unit<H: Hasher>(
+        Ok(Some(round)) => round,
+        Ok(None) => {
+            warn!(target: "AlephBFT-creator", "None starting round provided. Exiting.");
+            return;
+        }
+        Err(e) => {
+            error!(target: "AlephBFT-creator", "Starting round not provided: {}", e);
+            return;
+        }
+    };
+
+    if let Err(err) = run_creator(conf, io, starting_round).await {
+        match err {
+            CreatorError::OutChannelClosed(e) => {
+                warn!(target: "AlephBFT-creator", "Notification send error: {}. Exiting.", e)
+            }
+            CreatorError::ParentsChannelClosed => {
+                debug!(target: "AlephBFT-creator", "Incoming parent channel closed, exiting.")
+            }
+        }
+    }
+}
+
+async fn run_creator<H: Hasher>(
+    conf: Config,
+    io: &mut IO<H>,
+    starting_round: Round,
+) -> anyhow::Result<(), CreatorError> {
     let Config {
         node_id,
         n_members,
         create_lag,
         max_round,
     } = conf;
     let mut creator = Creator::new(node_id, n_members);
-    let IO {
-        mut incoming_parents,
-        outgoing_units,
-    } = io;
-
-    let starting_round = futures::select! {
-        maybe_round =  starting_round => match maybe_round {
-            Ok(Some(round)) => round,
-            Ok(None) => {
-                warn!(target: "AlephBFT-creator", "None starting round provided. Exiting.");
-                return;
-            }
-            Err(e) => {
-                error!(target: "AlephBFT-creator", "Starting round not provided: {}", e);
-                return;
-            }
-        },
-        _ = &mut terminator.get_exit() => {
-            terminator.terminate_sync().await;
-            return;
-        },
-    };
+    let incoming_parents = &mut io.incoming_parents;
+    let outgoing_units = &io.outgoing_units;
 
     debug!(target: "AlephBFT-creator", "Creator starting from round {}", starting_round);
     for round in starting_round..max_round {
         // Skip waiting if someone created a unit of a higher round.
         // In such a case at least 2/3 nodes created units from this round so we aren't skipping a
         // delay we should observe.
-        let ignore_delay = creator.current_round() > round;
-        let (unit, parent_hashes) = match create_unit(
-            round,
-            &mut creator,
-            &create_lag,
-            ignore_delay,
-            &mut incoming_parents,
-            terminator.get_exit(),
-        )
-        .await
-        {
-            Ok((u, ph)) => (u, ph),
-            Err(_) => {
-                terminator.terminate_sync().await;
-                return;
-            }
-        };
+        // NOTE: even we've observed a unit from a higher round, our own unit from previous round
+        // might not be yet added to `creator`. We still might need to wait for its arrival.
+        let should_delay = !(creator.current_round() > round);
+        if should_delay {
```

Comment:
The if could have the condition directly in it, without the additional variable.

I'm also pretty sure you shouldn't add the "NOTE" now, nothing in the code currently suggests we expect to be able to create a unit immediately, just that we will not be observing the artificial delay.

At the same time the comment above might require a slight rework – now we are waiting unless we are behind. Hm, not sure I am expressing this clearly, if it looks good to you leave it as-is.

Reply:
IMHO it's more human-friendly :) to read a bit more verbose version, but I am ok with removing it. Maybe just change that variable to `skip_delay` and leave the comment intact.

Reply:
In general I agree with more descriptions, but in this case this "debunked" a misconception the reader was unlikely to develop in the current version of the code, so it would confuse them until they figured out why it might have been an issue in the first place. So, increasing cognitive load instead of decreasing it.

I'm less sure about the variable, I see why it might be nice to have such semi-useless variables for readability sometimes and this is a borderline case I guess.